### PR TITLE
Implemented expectedGasLimit from spec

### PR DESCRIPTION
**Pull request by rolfyone**
Made a more complete test suite on the FN to show boundaries.

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java [hunk 1]
```diff
@@ -122,20 +122,21 @@ public void validateBuilderBid(
...
+      eventLogger.builderBidNotHonouringGasLimit(
+          parentGasLimit, proposedGasLimit, preferredGasLimit);
     }
+  }
 
-    if (preferredGasLimit.isLessThan(parentGasLimit)
-        && proposedGasLimit.isLessThan(parentGasLimit)) {
-      return;
+  static UInt64 expectedGasLimit(final UInt64 parentGasLimit, final UInt64 targetGasLimit) {
+    final UInt64 maxGasLimitDifference = parentGasLimit.dividedBy(1024L).minusMinZero(1);
```
Comment (jtraglia) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java:133]: Why is the `.minusMinZero(1)` here?

Ah. It's because it asserts that it's strictly less than or greater than the value.

image

For example, given that

* `parent_gas_limit == 1024_000`
* `parent_gas_limit // 1024 == 1000`

`gas_limit` must be less than `1024_000 + 1000` (1025_000), so the max value is 1024_999.

Comment (rolfyone) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java:133]: i'll try to find the original spec entry, i was referencing a PR from prysm, which did the -1...

File: ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorTest.java [hunk 2]
```diff
@@ -184,27 +200,44 @@ void shouldNotLogEventIfGasLimitStaysTheSame() throws BuilderBidValidationExcept
...
+        // down inside bounds
+        Arguments.of(1024_000L, 1023_500L, 1023_500L),
+        // down outside bounds - results in a partial move
+        Arguments.of(36_000_000L, 35_000_000L, 35_964_845L),
+        Arguments.of(1024_000L, 1020_000L, 1023_001L),
+        // increase outside bounds - results in a partial move
+        Arguments.of(1024_000L, 1025_000L, 1024_999L),
+        Arguments.of(30_000_000L, 36_000_000L, 30_029_295L),
+        // increase inside bounds
+        Arguments.of(1024_000L, 1024_500L, 1024_500L));
```
Comment (rolfyone) [inline ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorTest.java:240]: @jtraglia so the args are parent, target, expected output...

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java (hunk 1 above)
Comment (rolfyone) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java:133]: @jtraglia The test below does demonstrate the bounds of this function also... i think you've explained exactly why the minus 1...

Comment (rolfyone) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java:133]: i think given the thread and the test match, i'll mark this resolved but feel free to ping me if this is an incorrect assertion

**Review by tbenr — APPROVED** [review]
LGTM

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java (hunk 1 above)
Comment (nflaig) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/BuilderBidValidatorImpl.java:133]: > i'll try to find the original spec entry, i was referencing a PR from prysm, which did the -1...

so I ended up here while looking for this in the spec, only to notice it actually is not part of any spec but only documented in EIP-1559...

@jtraglia should we add this to the bellatrix builder spec?
